### PR TITLE
Move precision landing from Navigator into Flight Tasks

### PR DESCRIPTION
**Pull request by potaito**
@ndepal FYI

**Describe problem solved by this pull request**
Partially solves https://github.com/PX4/PX4-Autopilot/issues/14320

Precision landing is congesting navigator when it's actually better suited to be a flight task. The navigator based precision landing implementation is quite tedious as it's constantly converting between local and global coordinates. This is a limitation of Navigator. By moving it to the flight task library, precision landing can directly interface with the local position and velocity setpoints. It's also easier to create new search patterns in the future for those who need it. Last but not least the implementation as a FlightTask should make it possible to use precision landing without a global position (GPS), which is a requirement in the previous implementation. I have not tested precision landing without GPS, but in theory it's possible to use it with only a local frame.

**Describe your solution**
Precision landing was moved from navigator to the flight task library. It's running in parallel to navigator's RTL mode. Initially the precland flight task will listen to Navigator setpoints while Navigator thinks it's returning home. Once the drone either detects the landing target, or is above home and starts descending, the flight task will take over and land on the target instead on home. It it does not see the target, it will pause during Navigator's landing sequence, wait and see if it can still detect the target within a timeout.

The advantage of "hijacking" the navigator's RTL is, that all Navigator features and parameters are functional (return altitude, geofences etc.) until the drone enters the landing phase.

I tried to maintain the previous functionality of precland in Navigator as good as possible. One feature is now a bit different  though: If the target is lost while it's being approached, the drone will now continue with the original Navigator RTL until it sees the target again. Originally the drone would freeze mid-air and just wait for the target to re-appear. Despite this fact, the refactored implementation does not violate the flow described in the PX4 docs: [http://docs.px4.io/master/en/advanced_features/precland.html#precision-landing](http://docs.px4.io/master/en/advanced_features/precland.html#precision-landing
)

Precland Refactor

In summary, the drone will either land on the target, or it will land on home. Those are the only options. 

**Test data / coverage**
Testing is a bit tricky, as MAVSDK currently does not have the IR_beacon message implemented and it's therefore not possible to create SITL tests. What I did is send the land_target message from within PX4 and run manual tests.

----- 

The following log shows how RTL is executed and the drone lands on a target at (x,y) = (3,1.5)

image

source: https://logs.px4.io/plot_app?log=30855969-ae58-48a9-b06e-b33599a202eb

-----

Here we see the configuration where precision landing is required, but the target not visible. The drone will wait and "search" for the target where the time marker is located in the graphs. After the configurable timeout of 10 seconds is reached, the drone proceeds with the usual landing as instructed by Navigator:

image

source: https://logs.px4.io/plot_app?log=f8da0f50-69e3-40f2-a687-d0520e74a6db

**TODO**

- [ ] Fix flimsy land detection in simulation
- [ ] Prevent flight task from being activated over and over again between land detected and disarm. This happens with all flight tasks currently.
- [x] Implement switching to precland as part of the mission, where `nav_state == AUTO_MISSION`
- [x] Fix opportunistic landing in RTL: Currently precland witll switch to fallback right away when the beacon is not visible, instead of flying back to HOME first before abandoning search.

### CODE REVIEW

**Comment by potaito** [comment]
@MaEtUgR I noticed that the flight tasks are constantly being activated() between the drone touching the ground, i.e. land detected, and the actual disarm. It's not causing any issues other than the console is being spammed by the precision landing state machine. Are you aware of this and this a problem, or intended behaviour?

Also, land-detection is currently flimsy. I'm landing with the z-position setpoint set to NAN, therefore using only the z-velocity setpoint. I think this should be enough, but sometimes the simulated drone in jmavsim enters a never ending cycle of 
```
INFO  [commander] Landing detected	
INFO  [commander] Takeoff detected	
INFO  [commander] Landing detected	
INFO  [commander] Takeoff detected	
INFO  [commander] Landing detected	
INFO  [commander] Takeoff detected	
INFO  [commander] Landing detected	
```
Any thoughts on that? :sweat_smile:

File: src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+
+	case PrecLandState::Fallback:
+		return true;
+
+	default:
+		return false;
+	}
+}
+
+void FlightTaskAutoPrecisionLanding::slewrate(float &sp_x, float &sp_y)
```
Comment (potaito) [inline src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp:404]: This whole thing can probably be replaced by an AlphaFilter...

**Comment by dusan19** [comment]
didnt read the code yet, just have a quick question.
Compared to the RTL "hijacking" logic, what about auto missions that end with precision landing mission item, is that still supported? In that case this flight task has to run in parallel with navigators mission not RTL.

> Testing is a bit tricky, as MAVSDK currently does not have the IR_beacon message implemented and it's therefore not possible to create SITL tests. What I did is send the land_target message from within PX4 and run manual tests.

Btw I normally test precision landing in simulation using gazebo iris_irlock world, that one has an irlock model and reports landing_target_pose when it sees it.

**Comment by potaito** [comment]
Hey Dusan :blush: 

> didnt read the code yet, just have a quick question. Compared to the RTL "hijacking" logic, what about auto missions that end with precision landing mission item, is that still supported? In that case this flight task has to run in parallel with navigators mission not RTL.

Good point. I think it should work if the mission has an RTL element at the end. So from a Navigator / Mission perspective it will execute RTL, but ultimately it will not land at home and on the target instead.

> Btw I normally test precision landing in simulation using gazebo iris_irlock world, that one has an irlock model and reports landing_target_pose when it sees it.

Ah yes, thanks!

File: msg/precision_landing_status.msg [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# ORBIT_YAW_BEHAVIOUR
```
Comment (dagar) [inline msg/precision_landing_status.msg:1]: ORBIT_YAW_BEHAVIOUR

Comment (mcsauder) [reply msg/precision_landing_status.msg:1]: Hi @potaito , I's love to see this PR make it in.  Do you want to commit above suggestion and rebase on upstream?

Comment (potaito) [reply msg/precision_landing_status.msg:1]: Sure can do :+1: 
I was expecting more comments. This can't be the only thing that's not optimal :joy: 

As far as I know, PrecisionLanding is not used a lot by the community, and the folks who use it probably have their own implementation of it. So maybe there is someone who would be unhappy with this PR, but since the PrecisionLanding is then a flight task, customizing and changing it would be super simple.

Comment (potaito) [reply msg/precision_landing_status.msg]: Done, there were only minor conflicts when rebasing.

**Comment by potaito** [comment]
Tested with `make px4_sitl gazebo_iris_irlock` as suggested by @dusan19.
Landing on the target works when the RTL's home location is close to the tag, and precision landing also works if the RTL return path intersects the beacon, in which case it lands early and on the beacon :+1:

**Comment by julianoes** [comment]
@potaito I rebased and force-pushed this. I hope you don't mind. Would be nice to get it in.

**Comment by potaito** [comment]
> @potaito I rebased and force-pushed this. I hope you don't mind. Would be nice to get it in.

Thanks! No, not at all :)

> And you're saying MAVSDK should have support for https://mavlink.io/en/messages/common.html#LANDING_TARGET and then we could test this?

Yes! It's been on my TODO list, but I never created a new plugin in MAVSDK before. Basically I would like to have the same interface in MAVSDK as for the follow-me message `FOLLOW_TARGET`, which proved very useful.

**Comment by julianoes** [comment]
@potaito ok, would you use the position fields in the message as well?

And why do we need to send the position using mavsdk if there is a beacon already in Gazebo using the `_irlock` model? I guess I'm a bit confused how it all works together.

**Comment by julianoes** [comment]
> Currently only precision landing is using the LandingTargetEstimator and it is a waste to have it running in the background all the time. With this change the estimator is only started once the precision landing flight task is activated, and stopped when the flight task is deactivated.

I'm not sure that makes sense. Shouldn't the estimator always run and estimate the landing target and then once you switch to landing it already knows where the landing target is, if it is available?

**Comment by julianoes** [comment]
@potaito ok, I reverted your two last commits, because I think you want to estimate the landing target at all times, and then use it immediately when you're trying to land.

I then tested it with `make px4_sitl gazebo_iris_irlock` and realized I had to actually enable precision landing with the param, so I made that the default.

And, I added in the land mode, so for me now both RTL and Land mode work.

**Comment by julianoes** [comment]
@MaEtUgR would be great to get your review of the new flight task part.

**Comment by potaito** [comment]
> ok, I reverted your two last commits, because I think you want to estimate the landing target at all times, and then use it immediately when you're trying to land.

That was on purpose :sweat_smile: 
But down the line I want to have the same estimator for follow-me and precision landing, and use the same uORB interface for both. Whether we chase or land on a target should not make a huge difference. So I guess it's fine if you want to have the estimator running all the time. For now I just didn't see a benefit, because it initializes within one second when measurements start coming in. 

 
> I then tested it with `make px4_sitl gazebo_iris_irlock` and realized I had to actually enable precision landing with the param, so I made that the default.

Thanks!

> And, I added in the land mode, so for me now both RTL and Land mode work.

Thanks also for that. It's probably expected that precision landing also works in LAND flight mode, especially if it works in RTL :+1:

**Comment by julianoes** [comment]
> That was on purpose :sweat_smile:

I know I know, feel free to remove my commits again if you disagree. I just sort of worked my way through and it didn't actually work with these changes. I think in general we try to avoid allocating and de-allocating once flying, so I'm not sure if we should start and stop work queues while in-air.

**Comment by potaito** [comment]
> > That was on purpose sweat_smile
> 
> I know I know, feel free to remove my commits again if you disagree. Sorry that I was a bit harsh. I just sort of worked my way through and it didn't actually work with these changes. I think in general we try to avoid allocating and de-allocating once flying, so I'm not sure if we should start and stop work queues while in-air.

Interesting. I ended up doing this because @bkueng made the suggestion when I was working on moving follow-me from navigator to flight tasks. I.e. start the follow-me estimator when the follow-me flight mode is entered. I don't have an opinion as to what's better on a low level OS perspective

**Comment by julianoes** [comment]
@potaito yer I might be wrong. But even for follow me, wouldn't you want to have the follower ready and tracking when you switch to it? You wouldn't want to wait 1s for it to converge. You basically would want to know if it is tracking before you switch, no?

**Comment by bkueng** [comment]
> I think in general we try to avoid allocating and de-allocating once flying, so I'm not sure if we should start and stop work queues while in-air.

It won't (re)allocate as it runs on the same wq. The run condition is something to think about though.

**Comment by potaito** [comment]
> @potaito yer I might be wrong. But even for follow me, wouldn't you want to have the follower ready and tracking when you switch to it? You wouldn't want to wait 1s for it to converge. You basically would want to know if it is tracking before you switch, no?

From an estimation perspective yes, absolutely. You'd want your estimator to be running for as long as possible before you start using it. An argument against would be that the estimator is wasting CPU cycles while it's not used. Then again it will be wasting the same cycles when precision landing is enabled, since no other task is kicked out of the work queue. So in effect we would not be saving anything :thinking: 

Both approaches are fine for me. As I wrote before, I'm certain I'll be merging follow-me and precision-landing at some point in the future such that the same estimator will be used for both, and that PX4 can land on the target it was following. For this feature to work, the estimator will need to keep running when switching from follow-me to precision landing and vice versa. I'm not sure how to implement this in an elegant fashion using the work queue. Instead of having the flight tasks start / stop the estimator job on the work queue, we could instead have the flight task manager perform this action and keep the estimator running when switching between tasks that both need it. If this approach works, then I have no bigger arguments at all in favour or against running the estimator on the work queue.

@julianoes one argument I can think of: When using the work queue, you don't need to worry about the user "enabling" the estimator when he/she wants to use follow-me or precision landing, as it will be started on demand. Currently the estimator is started inside of the airframe configuration: https://github.com/PX4/PX4-Autopilot/blob/54438140e12b44654b08e890399f52d564d89874/ROMFS/px4fmu_common/init.d-posix/airframes/1011_iris_irlock#L15, which I find cumbersome for new users. Alternatively we put `landing_target_estimator start` into the general `rcS` script and enable / disable the estimator via parameter. But that again requires a manual step by the user. Having it started on the work queue by the flight task itself eliminates the need for any user configuration.

So in the end it's "Ease of use" versus "Estimator already converged". I would vote for ease of use, since there are messages on Slack asking how to use follow-me or precision landing every 1-2 weeks.

**Comment by dusan19** [comment]
```
} else if ((_vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_RTL ||
		    _vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LAND)
```
This is the condition to switch to precland flight task. It seems to me that it wont switch to precland when precland is requested as a part of the mission (for mission item of type MAV_CMD_NAV_LAND with a precland parameter set (param 2)), In that case _vehicle_status_sub.get().nav_state is of type AUTO_MISSION which isnt handled here. Am I missing something?

I have another question. I dont understand the RTL logic now, im sure you tested and it works, but would like some clarification for my own sake. The condition above switches the flight task to precland when nav_state is RTL, but does not necessarily check for the landing phase of RTL. So if the vehicle is not above the beacon and RTL is triggered, then how does running the precland task work? Wouldnt it just start searching for the beacon instead of climbing and going to the home position first?

**Comment by potaito** [comment]
@dusan19 
> This is the condition to switch to precland flight task. It seems to me that it wont switch to precland when precland is requested as a part of the mission (for mission item of type MAV_CMD_NAV_LAND with a precland parameter set (param 2)), In that case _vehicle_status_sub.get().nav_state is of type AUTO_MISSION which isnt handled here. Am I missing something?

Nope, you are right. It's not handled. I was not aware of this command flow. :+1: 
Thanks for your feedback!

> I have another question. I dont understand the RTL logic now, im sure you tested and it works, but would like some clarification for my own sake. The condition above switches the flight task to precland when nav_state is RTL, but does not necessarily check for the landing phase of RTL. 

Yeah it's a bit confusing, I have to admit that. Here's how it works:
The flight task is activated and follows navigator's position setpoint triplet: https://github.com/PX4/PX4-Autopilot/pull/18973/files#diff-89d61979c7d58eb43d987eaaf69c511d11a7eee9bd04b990f85ea3fd8e029ed0R122
 
The flight task stops listening to navigator / the position setpoint triplet once navigators begins to land: https://github.com/PX4/PX4-Autopilot/pull/18973/files#diff-89d61979c7d58eb43d987eaaf69c511d11a7eee9bd04b990f85ea3fd8e029ed0R133-R136

I wanted to keep this refactor as close as possible to the existing implementation, which is why the logic is now a bit of a mishmash. 

> So if the vehicle is not above the beacon and RTL is triggered, then how does running the precland task work? Wouldnt it just start searching for the beacon instead of climbing and going to the home position first?

See the start block of the flight task: https://github.com/PX4/PX4-Autopilot/pull/18973/files#diff-89d61979c7d58eb43d987eaaf69c511d11a7eee9bd04b990f85ea3fd8e029ed0R124-R137

In other words: If the beacon is visible, it will approach the beacon horizontally. If it's not visible and opportunistic precland is enabled, it will switch to fallback and ignore precland. Third option, and this is the most common one I think, if the beacon is not visible not precland is required, it will listen to Navigator's position setpoint triplet, which is performing a common RTL, and then it will take over in the land phase.

**Comment by julianoes** [comment]
> @julianoes one argument I can think of: When using the work queue, you don't need to worry about the user "enabling" the estimator when he/she wants to use follow-me or precision landing, as it will be started on demand. Currently the estimator is started inside of the airframe configuration:

The way this should work is that the estimator is automatically started when the param to enable it is set to `Opportunistic` or `Required`, so in the startup script it should be:

```
if param greater RTL_PLD_MD 0
then
    landing_target_estimator start
fi
```
And that param then needs to be marked as `@reboot_required`.

The one argument I have to always run the estimator is that you want to know the CPU required and not be surprised mid-flight when it suddenly goes up. So, I'd say, you might as well always have it running.

**Comment by potaito** [comment]
> > @julianoes one argument I can think of: When using the work queue, you don't need to worry about the user "enabling" the estimator when he/she wants to use follow-me or precision landing, as it will be started on demand. Currently the estimator is started inside of the airframe configuration:
> 
> The way this should work is that the estimator is automatically started when the param to enable it is set to `Opportunistic` or `Required`, so in the startup script it should be:
> 
> ```
> if param greater RTL_PLD_MD 0
> then
>     landing_target_estimator start
> fi
> ```
> 
> And that param then needs to be marked as `@reboot_required`.
> 
> The one argument I have to always run the estimator is that you want to know the CPU required and not be surprised mid-flight when it suddenly goes up. So, I'd say, you might as well always have it running.

Fair enough, you are right!

**Comment by potaito** [comment]
Thanks @dusan19, you are right again! I'll fix this and the mission logic that is not being handled.

**Comment by potaito** [comment]
@julianoes

> The way this should work is that the estimator is automatically started when the param to enable it is set to `Opportunistic` or `Required`, so in the startup script it should be:
> 
> if param greater RTL_PLD_MD 0
> then
>     landing_target_estimator start
> fi
> ```

What should happen if `RTL_PLD_MD = 0`, but then the user activates the "Precision Landing" mode in QGC's flight mode dropdown list? :sweat_smile: 

The estimator wouldn't be running, so the flight mode wouldn't work.

File: src/modules/flight_mode_manager/FlightModeManager.cpp [hunk 3]
```diff
@@ -215,6 +215,29 @@ void FlightModeManager::start_flight_task()
 			_task_failure_count = 0;
 		}
 
+		// Take-over landing from navigator if precision landing is enabled
+
+	} else if ((_vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_RTL ||
+		    _vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LAND)
+		   && _param_rtl_pld_md.get() > 0) {
```
Comment (dakejahl) [inline src/modules/flight_mode_manager/FlightModeManager.cpp:222]: Does using this flag for both RTL and Land have any unintended consequences? What if you want required precision land for RTL but not for land? I can definitely see a case where you might have a landing target back home you are required to land on, but during a flight you may want to trigger a land for safety/other reasons. In this case the land logic would be broken.

I see Dusan already pointed this out but not all AUTO_LANDs will want to be precision lands so we need to somehow pass the state of `land_precision` [in navigator ](https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/mavlink/mavlink_mission.cpp#L1384-L1389)to this flight task to handle Mission NAV_CMD_LAND precision lands.

Comment (dakejahl) [reply src/modules/flight_mode_manager/FlightModeManager.cpp:222]: hehe just found this https://docs.google.com/document/d/1N59JHFqL_E9X4b-U0CzxW_eydqXUHZMxAH_FTD30ilk/edit

>Refactor Navigator into having one FlightMode per navigation state

I think this implies any behavior that is shared between modes and mission items need to be handled as a FlightTask (eg RTL, Land, Precision Land, Takeoff, Loiter, etc). Am I understanding this correctly? Maybe we can publish a new uORB topic `navigation_mode` from navigator which the `FlightModeManager` listens to and switches to the correct task by checking

For mission items
```
_vehicle_control_mode_sub.get().flag_control_auto_enabled && (_nav_mode_sub.get().navigation_mode == MODE_OF_INTEREST)
```

For modes
```
_vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_*
```

This will kill off nearly all of the MissionBlock classes in navigator and consolidate control code to a single place (FlightTasks)

**Comment by julianoes** [comment]
@potaito my only (bad) idea is to always run the estimator :thinking:.

File: src/modules/flight_mode_manager/FlightModeManager.cpp (hunk 3 above)
Comment (potaito) [reply src/modules/flight_mode_manager/FlightModeManager.cpp:222]: > What if you want required precision land for RTL but not for land
That feature does not exist yet :sweat_smile: 

Assuming I manage to make it work with missions, you would then have to set `RTL_PLD_MD=0` to disable precision landing, and then specifically enable it for a mission item. That's the only way I can think of that does not involve yet another parameter for precision LAND besides precision RTL.

Comment (potaito) [reply src/modules/flight_mode_manager/FlightModeManager.cpp:222]: > I think allowing precision land for the land command (NAVIGATION_STATE_AUTO_LAND) is a good idea but it should be a separate parameter to enable it.

I'm just worried that we end up creating too many parameters for something simple. 

There's also the DESCEND flight mode, which does not control position, but only velocity. Usually used when GPS is lost mid-flight. Theoretically DESCEND could still perform precision landing as long as the target is visible, even without GPS.

So I think what all of this tells me is that I'll now make all existing precland use cases work (precland as part of RTL, flight mode or mission) and then in a second iteration actually improve the architecture as you suggested.

**Comment by potaito** [comment]
@dusan19 @dakejahl 
In [the last commit](https://github.com/PX4/PX4-Autopilot/pull/18973/commits/571cb999efea6ffebed5f041700537b2c97ea3b6) I broke out information about the sub-nav-mode during missions, so that FlightModeManager knows when to engage precland during a mission.

With this commit in place precision landing can be engaged as follows:
- Activate the precland flight mode, for example with the QGC drop down menu. This will be translated to a vehicle_command in PX4
- Use `commander mode auto:precland` in the nuttx shell
- Create a mission with a land item in it where the precland parameter is set to opportunistic or required
- Set the PX4 parameter `RTL_PLD_MD` to "opportunistic" or "required" and then enter RTL or LAND flight mode.

-----

@dagar @julianoes and everyone else who cares, we will need to talk at some point :sweat_smile: 

This whole exercise of moving precland out of navigator into a fight task is raising a lot of questions that need to be solved in general for PX4. For example:
- What is / should be the role of navigator in general? I think we all agree that currently it is doing too much. Should it be a state machine? Or should it only be an array of mission items? 
- How can flight modes such as precland, orbit or follow-me become part of a mission?
- If we allow custom flight modes as flight tasks to be used in missions, the flight tasks will have to contain the criteria that determine when the task is complete. For example, if the vehicle is supposed to fly orbit for three resolutions, then the orbit flight task should track this and report when it's done.
- What are AUTO modes? Is AUTO = mission? Or is AUTO a superset that contains mission along other modes?

The prototype solution I propose here is that Navigator can be in NAVIGATION_STATE_AUTO_MISSION and then publish it's own sub mode during a mission, which in this case is a precision land. I think this is in line with what developers before me had in mind as well. Commander has a notion of main modes and sub modes, also for precision landing:

https://github.com/PX4/PX4-Autopilot/blob/fe26ee244d5ecb8e6610b6146dc6102cf3b691a4/src/modules/commander/Commander.cpp#L398-L401

But it's used differently than what I have in mind with mission as main mode, and precland as submode. There is a general convolution of the terms AUTO and MISSION in PX4 that we need to resolve.

File: src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp (hunk 1 above)
Comment (potaito) [reply src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp:417]: There is actually a SlewRate library in PX4. But that's for another day.

File: src/modules/flight_mode_manager/FlightModeManager.cpp (hunk 3 above)
Comment (potaito) [reply src/modules/flight_mode_manager/FlightModeManager.cpp]: @dakejahl 
> What if you want required precision land for RTL but not for land

I thought about it, and in my opinion we should get rid of the `RTL_PLD_MD` parameter in the first place. Otherwise we would have to also add a parameter for precision land mode `LND_PLD_MD` or whatever, and perhaps for other modes as well where you wish to have the precision-target being used somehow.

Down the line I would just make them separate flight modes. Then we would have
- RTL
- PrecisionRTL
- Land
- PrecisionLand
- whatever else you can think of

This would also reduce the number of code paths for how one can end up in PrecisionLand, which right now is very confusing.

**Comment by potaito** [comment]
With the recent changes, I fixed the behaviour that @dusan19 mentioned where a combination of opportunistic precision landing + RTL + the beacon not being visible would result in an instant fallback land, rather than executing a generic RTL.

I would like to merge this as is now, since I'll refactor the flight task anyway. This PR is intended as the architectural basis for re-inventing the precision landing as a flight task.

I tried to create a diagram of the information flow after this refactor. All depicted connections to FlightModeManager are new:

Precland Refactor

File: src/modules/flight_mode_manager/FlightModeManager.cpp (hunk 3 above)
Comment (dakejahl) [reply src/modules/flight_mode_manager/FlightModeManager.cpp]: makes sense, also makes it more clear about what a mode is and how you get into it

File: src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+void
+FlightTaskAutoPrecisionLanding::run_state_descend_above_target()
+{
+	// Overwrite Auto setpoints in order to descend above target
+	_position_setpoint(0) = _target_pose.x_abs;
+	_position_setpoint(1) = _target_pose.y_abs;
+	_position_setpoint(2) = NAN;
+	_velocity_setpoint(0) = 0;
+	_velocity_setpoint(1) = 0;
+	_velocity_setpoint(2) = _param_mpc_land_speed.get();
```
Comment (dakejahl) [inline src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp:200]: This should use the same logic here for the different land speeds and heights
https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp#L236-L245

**Review by dakejahl — DISMISSED** [review]

**Comment by junwoo091400** [comment]
The `mavlink_mission.cpp` part of handling the flag setting for precision landing is currently not implemented in this PR (It is included in the diagram) :thinking:

**Comment by junwoo091400** [comment]
Oops no, this was already implemented before the PR, sorry for the confusion!

**Comment by potaito** [comment]
Yep:

https://github.com/PX4/PX4-Autopilot/blob/132a07f53cddc61bca3c241eda1a9e148c696d43/src/modules/mavlink/mavlink_mission.cpp#L1391-L1396

File: src/modules/flight_mode_manager/FlightModeManager.cpp [hunk 5]
```diff
@@ -169,6 +170,16 @@ void FlightModeManager::start_flight_task()
...
 
+	const bool land_should_be_precland = (_vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_RTL
+					      ||
+					      _vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LAND)
+					     && _param_rtl_pld_md.get() > 0;
+	const bool precland_mission_item_active = _vehicle_status_sub.get().nav_state ==
+			vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION &&
+			_navigator_mission_item_sub.get().nav_sub_cmd == navigator_mission_item_s::WORK_ITEM_TYPE_PRECISION_LAND;
+	const bool precland_flight_mode = _vehicle_status_sub.get().nav_state ==
+					  vehicle_status_s::NAVIGATION_STATE_AUTO_PRECLAND;
```
[the remark this answers is not in the document]
Comment (potaito) [reply src/modules/flight_mode_manager/FlightModeManager.cpp:181]: You're right. It should probably be hidden away in some separate function. But FlightModeManager is the right place to contain this logic. It's not making a decision, it's just figuring out what navigator and commander already decided.

File: src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+FlightTaskAutoPrecisionLanding::run_state_touching_down()
+{
+	// Overwrite Auto setpoints in order to land at target's last known location
+	_position_setpoint(0) = _landing_target_pose.x_abs;
+	_position_setpoint(1) = _landing_target_pose.y_abs;
+	_position_setpoint(2) = NAN;
+
+	_velocity_setpoint(0) = 0;
+	_velocity_setpoint(1) = 0;
+	_velocity_setpoint(2) = _param_mpc_land_speed.get();
```
Comment (dakejahl) [inline src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp:230]: Use land speed parameters

**Review by dakejahl — COMMENTED** [review]
Do you need me to test this in sim? I'm on Ubuntu 20.04.

**Comment by dakejahl** [comment]
I just tested this briefly in sim and the drone flew away during Search :joy: 

It also spams
```
[FlightTaskAutoPrecisionLanding] Precland: switching to horizontal approach
```
after touch down on landing.

I will leave it be on testing for now while you're still working on it.

File: src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+}
+
+void
+FlightTaskAutoPrecisionLanding::run_state_search()
+{
+	// Overwrite Auto setpoints in order to hover at search altitude
+	_position_setpoint = _target;
+	_position_setpoint(2) = _sub_home_position.get().z - _param_pld_srch_alt.get();
```
[the remark this answers is not in the document]
Comment (potaito) [reply src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp:238]: Yeah there's a separate PR to fix that in navigator asap: https://github.com/PX4/PX4-Autopilot/pull/21260

File: src/modules/flight_mode_manager/tasks/PrecisionLanding/FlightTaskPrecisionLanding.cpp [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+		break;
+
+	case PRECLAND_STATE::MOVE_TO_SEARCH_ALTITUDE:
+	case PRECLAND_STATE::SEARCHING_TARGET:
+	case PRECLAND_STATE::MOVING_ABOVE_TARGET:
+
+		// TODO: Should use current altitude when state was entered as the altitude setpoint
+		if (PX4_ISFINITE(_landing_target_pose.z_abs)) {
+			// For safety reasons take whichever reference altitude is higher up (lower in NED)
+			_position_setpoint(2) = math::min(_sub_home_position.get().z, _landing_target_pose.z_abs) - search_rel_altitude;
```
Comment (akkawimo) [inline src/modules/flight_mode_manager/tasks/PrecisionLanding/FlightTaskPrecisionLanding.cpp:135]: Does this account for the case where home/takeoff point is above the landing tag?
